src/main/java/com/gtnewhorizon/structurelib/structure/StructureDefinition.java

```
@@ -246,4 +331,61 @@ public boolean isContainedInStructure(String name, int offsetA, int offsetB, int
         if (occupiedSpace == null) throw new NoSuchElementException(name);
         return occupiedSpace.contains(new Vec3Impl(offsetA, offsetB, offsetC));
     }
+
+    @Override
+    public Position<StructureDefinitionCoords> getControllerPosition(String piece) {
+        return controllers.get(piece);
+    }
+
+    @Override
+    public StructureDefinitionCoords getCoordinateSystem(String piece) {
+        var controller = controllers.get(piece);
+
+        if (controller == null) {
+            throw new IllegalArgumentException("Invalid piece: " + piece);
+        }
+
+        return new StructureDefinitionCoords(controller.x, controller.y, controller.z);
+    }
+
+    @Override
+    public Position<StructureDefinitionCoords> getSocket(String piece, char socket) {
+        var piecePositions = socketLocations.get(piece);
+
+        if (piecePositions == null) {
+            throw new IllegalArgumentException("Invalid piece: " + piece);
```

piece might exist but have no socket. throwing an error would likely cause some headache in downstream repos that needs to introspect a structure e.g. blockrenderer

---

This is meant to be a "assert one and only one socket" type of method. `getAllSockets` only throws when the piece is missing, but I consider that a logic error regardless of the usage. If something is dynamically scanning for sockets, it should check if the piece exists first (though it doesn't look like that method exists, should I add it?)

---

doesn't getAllSockets throw when piece have no socket either? you did not eagerly create an entry in socketLocations for every piece. you did so only when it has a socket

---

No, multimaps return an empty collection for keys that don't exist. Since it's a ListMultimap it'll return an empty list.

---

> private final Map<String, ListMultimap<Character, Position<StructureDefinitionCoords>>> socketLocations = new HashMap<>();

socketLocations is a HashMap

---

The data is roughly `{piece name: {socket char: Position[]}}`. The outer piece -> socket map is a hash map, but the socket char -> position map is a multimap. I know it's an ugly unreadable type but I can't think of a way to improve it.

---

I'm not sure how to convey this better but let's just say this piece of code will throw with an exception saying main is invalid

```java
builder().addShape("main", new String[][] {{"~"}}).build().getAllSockets("main")
```

---

I figured out what you meant, it should be good now. I added some tests to validate that it's correct.